core/keystore/src/lib.rs

```diff
@@ -14,185 +14,180 @@
-		match self {
-			Error::Io(ref err) => Some(err),
-			Error::Json(ref err) => Some(err),
-			_ => None,
-		}
-	}
+/// Handler for `Store` events.
+trait StoreHandler: Send {
+	/// Called when a new session key is added to the store.
+	fn on_set_session_key(&self, pair: &[u8]);
 }
 
-/// Key store.
+/// Store for session keys.
+#[derive(Clone)]
 pub struct Store {
-	path: PathBuf,
-	additional: HashMap<(KeyTypeId, Vec<u8>), Vec<u8>>,
+	handlers: Arc<Mutex<Vec<(KeyTypeId, Box<dyn StoreHandler>)>>>,
 }
 
 impl Store {
-	/// Create a new store at the given path.
-	pub fn open(path: PathBuf) -> Result<Self> {
-		fs::create_dir_all(&path)?;
-		Ok(Store { path, additional: HashMap::new() })
+	/// Create a new store.
+	pub fn new() -> Self {
+		Self {
+			handlers: Default::default(),
+		}
 	}
 
-	fn get_pair<TPair: Pair>(&self, public: &TPair::Public) -> Result<Option<TPair>> {
-		let key = (TPair::KEY_TYPE, public.to_raw_vec());
-		if let Some(bytes) = self.additional.get(&key) {
-			let pair = TPair::from_seed_slice(bytes)
-				.map_err(|_| Error::InvalidSeed)?;
-			return Ok(Some(pair));
+	/// Adds a session key pair.
+	pub fn set_key<TPair: Pair>(&self, pair: &TPair) {
+		let value = pair.to_raw_vec();
+		let handlers = self.handlers.lock();
+		for (key_type, handler) in handlers.iter() {
+			if *key_type == TPair::KEY_TYPE {
+				handler.on_set_session_key(&value[..]);
+			}
 		}
-		Ok(None)
 	}
 
-	fn insert_pair<TPair: Pair>(&mut self, pair: &TPair) {
-		let key = (TPair::KEY_TYPE, pair.public().to_raw_vec());
-		self.additional.insert(key, pair.to_raw_vec());
+	/// Adds a session key pair from a seed. Used for testing.
+	pub fn set_key_from_seed<TPair: Pair>(&self, seed: &str) -> Result<()> {
+		let pair = TPair::from_string(seed, None)
+			.map_err(|_| Error::InvalidSeed)?;
+		self.set_key(&pair);
+		Ok(())
 	}
 
-	/// Generate a new key, placing it into the store.
-	pub fn generate<TPair: Pair>(&self, password: &str) -> Result<TPair> {
-		let (pair, phrase, _) = TPair::generate_with_phrase(Some(password));
-		let mut file = File::create(self.key_file_path::<TPair>(&pair.public()))?;
-		::serde_json::to_writer(&file, &phrase)?;
-		file.flush()?;
-		Ok(pair)
+	/// Creates a local session store.
+	pub fn create_local_store<TPair>(&self) -> LocalStore<TPair>
+	where
+		TPair: Pair + Send + Clone,
+		TPair::Public: Send + Clone,
+	{
+		let session_store = LocalStore::new();
+		self.handlers.lock().push((TPair::KEY_TYPE, Box::new(session_store.clone())));
+		session_store
 	}
+}
 
-	/// Create a new key from seed. Do not place it into the store.
-	pub fn generate_from_seed<TPair: Pair>(&mut self, seed: &str) -> Result<TPair> {
-		let pair = TPair::from_string(seed, None)
-			.ok().ok_or(Error::InvalidSeed)?;
-		self.insert_pair(&pair);
-		Ok(pair)
-	}
+/// A local session store supporting a single crypto algorithm.
+#[derive(Clone)]
+pub struct LocalStore<TPair: Pair> {
+	keys: Arc<Mutex<HashMap<TPair::Public, (TPair, Instant)>>>,
```

Comment:
I'd rather move `Arc` out of the struct. Cloning an `Arc<LocalStore>` shows the intention better imho - you can also get rid of a `Box` that way.